File: forecast.py
import datetime

import requests

from encoding import (
    VERSION,
    RESOLUTION_HOURS,
    MODEL_BIT,
    ForecastMessage,
    Period,
)

OPENMETEO_MODELS: dict[str, str] = {
    "ECMWF": "ecmwf_ifs025",
    "GFS": "gfs_seamless",
    "ICON": "icon_seamless",
}

LOCATION_COORDS: dict[str, tuple[float, float, str]] = {
    "upper": (63.0692, -151.0070, "America/Anchorage"),
    "airstrip": (62.900, -151.093, "America/Anchorage"),
}

# resolution index → hours to request from API (None = use noon daily rows)
RESOLUTION_TARGET_HOURS: dict[int, list[int] | None] = {
    0: None,
    1: [0, 12],
    2: [0, 6, 12, 18],
    3: [0, 3, 6, 9, 12, 15, 18, 21],
    4: list(range(24)),
}

_RESOLUTION_LABEL_TO_IDX: dict[str, int] = {
    "daily": 0, "24h": 0,
    "12h": 1,
    "6h": 2,
    "3h": 3,
    "1h": 4,
}

_MODEL_NAME_TO_BIT: dict[str, int] = {
    "ecmwf": MODEL_BIT["ECMWF"], "euro": MODEL_BIT["ECMWF"],
    "gfs": MODEL_BIT["GFS"],
    "icon": MODEL_BIT["ICON"],
}

_SURFACE_VARS = [
    "wind_speed_10m",
    "wind_direction_10m",
    "wind_gusts_10m",
    "precipitation_probability",
    "weather_code",
    "cloud_cover_mid",
    "freezing_level_height",
    "snowfall",
]
_PRESSURE_LEVELS = [500, 600, 700]
_PRESSURE_VAR_NAMES = ["temperature", "wind_speed", "wind_direction"]


def _deg_to_dir_idx(deg: float | None) -> int:
    if deg is None:
        return 0
    return round(deg / 45) % 8


def _round5(v: float | None) -> int:
    return round((v or 0) / 5) * 5


# ── API fetch helpers ─────────────────────────────────────────────────────────


def _fetch_hourly(
    model_key: str, n_days: int, lat: float, lon: float, tz: str
) -> tuple[dict, list[str]]:
    pressure_vars = [
        f"{v}_{l}hPa" for v in _PRESSURE_VAR_NAMES for l in _PRESSURE_LEVELS
    ]
    params = {
        "latitude": lat,
        "longitude": lon,
        "hourly": ",".join(_SURFACE_VARS + pressure_vars),
        "wind_speed_unit": "mph",
        "timezone": tz,
        "forecast_days": n_days,
        "models": OPENMETEO_MODELS[model_key],
    }
    resp = requests.get(
        "https://api.open-meteo.com/v1/forecast", params=params, timeout=15
    )
    resp.raise_for_status()
    h = resp.json()["hourly"]
    return h, h["time"]


def _build_row(h: dict, times: list[str], idx: int, snow_cm: float = 0.0) -> dict:
    row: dict = {
        "time": times[idx],
        "precip": h["precipitation_probability"][idx],
        "weathercode": h["weather_code"][idx],
        "cloudcover_mid": h["cloud_cover_mid"][idx],
        "freezing_level_m": h["freezing_level_height"][idx],
        "snow_cm": snow_cm,
    }
    for v in _PRESSURE_VAR_NAMES:
        for l in _PRESSURE_LEVELS:
            row[f"{v}_{l}hPa"] = h[f"{v}_{l}hPa"][idx]
    return row


def _noon_rows(
    model_key: str, n_days: int, lat: float, lon: float, tz: str
) -> list[dict]:
    """One row per day at noon local time, with cumulative daily snow."""
    h, times = _fetch_hourly(model_key, n_days, lat, lon, tz)
    snow_arr = h.get("snowfall", [])
    daily_snow: dict[str, float] = {}
    for i, t in enumerate(times):
        date = t[:10]
        daily_snow[date] = daily_snow.get(date, 0.0) + (snow_arr[i] or 0.0)
    rows = []
    for i, t in enumerate(times):
        if t.endswith("T12:00") and len(rows) < n_days:
            rows.append(_build_row(h, times, i, round(daily_snow.get(t[:10], 0.0), 1)))
    return rows


def _hour_rows(
    model_key: str,
    n_days: int,
    target_hours: list[int],
    lat: float,
    lon: float,
    tz: str,
) -> list[dict]:
    """Rows at specific hours across n_days."""
    h, times = _fetch_hourly(model_key, n_days, lat, lon, tz)
    snow_arr = h.get("snowfall", [])
    n_total = n_days * len(target_hours)
    rows = []
    for i, t in enumerate(times):
        hour = int(t[11:13])
        if hour in target_hours:
            rows.append(_build_row(h, times, i, snow_arr[i] or 0.0 if snow_arr else 0.0))
            if len(rows) == n_total:
                break
    return rows


# ── Row → Period ──────────────────────────────────────────────────────────────


def _to_full(r: dict, daily: bool = True) -> Period:
    fz_m = r.get("freezing_level_m") or 0
    snow_cm = r.get("snow_cm") or 0
    snow_in = round(snow_cm / 2.54) if daily else round(snow_cm / 0.254)
    return Period(
        weathercode=int(r.get("weathercode") or 0),
        precip=int(r.get("precip") or 0),
        freeze_ft=round(fz_m * 3.28084 / 1000) * 1000,
        snow_in=min(snow_in, 15),
        cloud_mid=int(r.get("cloudcover_mid") or 0),
        wind_700_mph=_round5(r.get("wind_speed_700hPa")),
        wind_700_dir=_deg_to_dir_idx(r.get("wind_direction_700hPa")),
        wind_500_mph=_round5(r.get("wind_speed_500hPa")),
        wind_500_dir=_deg_to_dir_idx(r.get("wind_direction_500hPa")),
        wind_600_mph=_round5(r.get("wind_speed_600hPa")),
        wind_600_dir=_deg_to_dir_idx(r.get("wind_direction_600hPa")),
    )


# ── Request parser ────────────────────────────────────────────────────────────


def parse_request(body: str) -> dict:
    """Parse 'upper 10d daily ecmwf,gfs' into fetch_forecast kwargs."""
    words = body.lower().strip().split()

    location_idx = 0
    days = 10
    resolution_idx = 0
    models_mask = 1  # ECMWF default

    for word in words:
        if word == "upper":
            location_idx = 0
        elif word == "airstrip":
            location_idx = 1
        elif word.endswith("d") and word[:-1].isdigit():
            days = max(1, min(10, int(word[:-1])))
        elif word in _RESOLUTION_LABEL_TO_IDX:
            resolution_idx = _RESOLUTION_LABEL_TO_IDX[word]
        else:
            model_parts = word.split(",")
            if any(m in _MODEL_NAME_TO_BIT for m in model_parts):
                mask = 0
                for m in model_parts:
                    if m in _MODEL_NAME_TO_BIT:
                        mask |= 1 << _MODEL_NAME_TO_BIT[m]
                if mask:
                    models_mask = mask

    return {
        "location_idx": location_idx,
        "days": days,
        "resolution_idx": resolution_idx,
        "models_mask": models_mask,
    }


# ── Forecast fetch ────────────────────────────────────────────────────────────


def fetch_forecast(
    location_idx: int = 0,
    days: int = 10,
    resolution_idx: int = 0,
    models_mask: int = 1,
) -> str:
    location_name = list(LOCATION_COORDS.keys())[location_idx]
    lat, lon, tz = LOCATION_COORDS[location_name]

    model_keys = [
        key
        for key, bit in [("ECMWF", 0), ("GFS", 1), ("ICON", 2)]
        if models_mask & (1 << bit)
    ] or ["ECMWF"]

    target_hours = RESOLUTION_TARGET_HOURS[resolution_idx]
    daily = resolution_idx == 0

    if daily:
        rows_per_model = [_noon_rows(m, days, lat, lon, tz) for m in model_keys]
    else:
        rows_per_model = [_hour_rows(m, days, target_hours, lat, lon, tz) for m in model_keys]

    first_time = rows_per_model[0][0]["time"]
    start_date = datetime.date.fromisoformat(first_time[:10])
    start_hour = int(first_time[11:13])

    return ForecastMessage(
        version=VERSION,
        location=location_idx,
        days=days,
        resolution=resolution_idx,
        models_mask=models_mask,
        month=start_date.month,
        day=start_date.day,
        hour=start_hour,
        periods=[[_to_full(r, daily=daily) for r in rows] for rows in rows_per_model],
    ).encode()


# ── __main__ ──────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    from encoding import CARDINALS, LOCATIONS, RESOLUTION_LABEL, models_from_mask

    test_requests = [
        "upper 10d daily ecmwf",
        "upper 5d 6h ecmwf,gfs",
        "airstrip 3d 12h ecmwf,gfs",
    ]
    for req_str in test_requests:
        print(f"\n{'=' * 60}")
        print(f"Request: {req_str!r}")
        params = parse_request(req_str)
        encoded = fetch_forecast(**params)
        decoded = ForecastMessage.decode(encoded)
        loc = LOCATIONS[decoded.location]
        res = RESOLUTION_LABEL[decoded.resolution]
        mods = models_from_mask(decoded.models_mask)
        print(f"Encoded ({len(encoded)} chars): {encoded}")
        print(f"Location: {loc}, Days: {decoded.days}, Res: {res}, Models: {mods}")
        print(f"Start: {decoded.start_datetime}")
        for m_idx, m_name in enumerate(mods):
            p = decoded.periods[m_idx][0]
            print(
                f"  {m_name} period 0: wc={p.weathercode} precip={p.precip}% "
                f"freeze={p.freeze_ft}ft snow={p.snow_in}in cloud={p.cloud_mid}% "
                f"700={p.wind_700_mph}mph {CARDINALS[p.wind_700_dir]}"
            )

File: packages/server/src/forecast.ts
import {
  VERSION,
  MODEL_BIT,
  type Period,
  type ForecastMessage,
  messageToString,
} from "@weather/protocol";

const OPENMETEO_MODELS: Record<string, string> = {
  ECMWF: "ecmwf_ifs",
  GFS: "gfs_seamless",
  ICON: "icon_seamless",
};

const LOCATION_COORDS: Record<string, [number, number, string]> = {
  upper: [63.135, -150.989, "America/Anchorage"],
  airstrip: [62.967, -151.057, "America/Anchorage"],
};

const RESOLUTION_TARGET_HOURS: Record<number, number[] | null> = {
  0: null,
  1: [0, 12],
  2: [0, 6, 12, 18],
  3: [0, 3, 6, 9, 12, 15, 18, 21],
  4: Array.from({ length: 24 }, (_, i) => i),
};

const RESOLUTION_LABEL_TO_IDX: Record<string, number> = {
  daily: 0,
  "24h": 0,
  "12h": 1,
  "6h": 2,
  "3h": 3,
  "1h": 4,
};

const MODEL_NAME_TO_BIT: Record<string, number> = {
  ecmwf: MODEL_BIT["ECMWF"],
  euro: MODEL_BIT["ECMWF"],
  gfs: MODEL_BIT["GFS"],
  icon: MODEL_BIT["ICON"],
};

const SURFACE_VARS = [
  "wind_speed_10m",
  "wind_direction_10m",
  "wind_gusts_10m",
  "precipitation_probability",
  "weather_code",
  "cloud_cover_mid",
  "freezing_level_height",
  "snowfall",
];
const PRESSURE_LEVELS = [500, 600, 700];
const PRESSURE_VAR_NAMES = ["temperature", "wind_speed", "wind_direction"];

function degToDirIdx(deg: number | null | undefined): number {
  if (deg == null) return 0;
  return Math.round(deg / 45) % 8;
}

function round5(v: number | null | undefined): number {
  return Math.round((v ?? 0) / 5) * 5;
}

interface HourlyData {
  time: string[];
  precipitation_probability: (number | null)[];
  weather_code: (number | null)[];
  cloud_cover_mid: (number | null)[];
  freezing_level_height: (number | null)[];
  snowfall: (number | null)[];
  [key: string]: unknown[];
}

interface Row {
  time: string;
  precip: number | null;
  weathercode: number | null;
  cloudcover_mid: number | null;
  freezing_level_m: number | null;
  snow_cm: number;
  wind_speed_500hPa: number | null;
  wind_direction_500hPa: number | null;
  wind_speed_600hPa: number | null;
  wind_direction_600hPa: number | null;
  wind_speed_700hPa: number | null;
  wind_direction_700hPa: number | null;
}

async function fetchHourly(
  modelKey: string,
  nDays: number,
  lat: number,
  lon: number,
  tz: string,
): Promise<[HourlyData, string[]]> {
  const pressureVars = PRESSURE_VAR_NAMES.flatMap((v) =>
    PRESSURE_LEVELS.map((l) => `${v}_${l}hPa`),
  );
  const params = new URLSearchParams({
    latitude: String(lat),
    longitude: String(lon),
    hourly: [...SURFACE_VARS, ...pressureVars].join(","),
    wind_speed_unit: "mph",
    timezone: tz,
    forecast_days: String(nDays),
    models: OPENMETEO_MODELS[modelKey],
  });
  const resp = await fetch(`https://api.open-meteo.com/v1/forecast?${params}`);
  if (!resp.ok) throw new Error(`Open-Meteo ${resp.status}: ${await resp.text()}`);
  const data = (await resp.json()) as { hourly: HourlyData };
  return [data.hourly, data.hourly.time];
}

function buildRow(h: HourlyData, times: string[], idx: number, snowCm: number): Row {
  return {
    time: times[idx],
    precip: h.precipitation_probability[idx],
    weathercode: h.weather_code[idx],
    cloudcover_mid: h.cloud_cover_mid[idx],
    freezing_level_m: h.freezing_level_height[idx],
    snow_cm: snowCm,
    wind_speed_500hPa: h["wind_speed_500hPa"][idx] as number | null,
    wind_direction_500hPa: h["wind_direction_500hPa"][idx] as number | null,
    wind_speed_600hPa: h["wind_speed_600hPa"][idx] as number | null,
    wind_direction_600hPa: h["wind_direction_600hPa"][idx] as number | null,
    wind_speed_700hPa: h["wind_speed_700hPa"][idx] as number | null,
    wind_direction_700hPa: h["wind_direction_700hPa"][idx] as number | null,
  };
}

async function noonRows(
  modelKey: string,
  nDays: number,
  lat: number,
  lon: number,
  tz: string,
): Promise<Row[]> {
  const [h, times] = await fetchHourly(modelKey, nDays, lat, lon, tz);
  const snowArr = h.snowfall;
  const dailySnow: Record<string, number> = {};
  for (let i = 0; i < times.length; i++) {
    const date = times[i].slice(0, 10);
    dailySnow[date] = (dailySnow[date] ?? 0) + ((snowArr[i] as number) || 0);
  }
  const rows: Row[] = [];
  for (let i = 0; i < times.length && rows.length < nDays; i++) {
    if (times[i].endsWith("T12:00")) {
      const date = times[i].slice(0, 10);
      rows.push(buildRow(h, times, i, Math.round((dailySnow[date] ?? 0) * 10) / 10));
    }
  }
  return rows;
}

async function hourRows(
  modelKey: string,
  nDays: number,
  targetHours: number[],
  lat: number,
  lon: number,
  tz: string,
): Promise<Row[]> {
  const [h, times] = await fetchHourly(modelKey, nDays, lat, lon, tz);
  const snowArr = h.snowfall;
  const nTotal = nDays * targetHours.length;
  const rows: Row[] = [];
  for (let i = 0; i < times.length && rows.length < nTotal; i++) {
    const hour = parseInt(times[i].slice(11, 13));
    if (targetHours.includes(hour)) {
      rows.push(buildRow(h, times, i, (snowArr[i] as number) || 0));
    }
  }
  return rows;
}

function toFullPeriod(r: Row, daily: boolean): Period {
  const fzM = r.freezing_level_m ?? 0;
  const snowIn = daily
    ? Math.round((r.snow_cm ?? 0) / 2.54)
    : Math.round((r.snow_cm ?? 0) / 0.254);
  return {
    weathercode: r.weathercode ?? 0,
    precip: r.precip ?? 0,
    freeze_ft: Math.round((fzM * 3.28084) / 1000) * 1000,
    snow_in: Math.min(snowIn, 15),
    cloud_mid: r.cloudcover_mid ?? 0,
    wind_500_mph: round5(r.wind_speed_500hPa),
    wind_500_dir: degToDirIdx(r.wind_direction_500hPa),
    wind_600_mph: round5(r.wind_speed_600hPa),
    wind_600_dir: degToDirIdx(r.wind_direction_600hPa),
    wind_700_mph: round5(r.wind_speed_700hPa),
    wind_700_dir: degToDirIdx(r.wind_direction_700hPa),
  };
}

export interface ForecastParams {
  locationIdx: number;
  lat?: number;
  lon?: number;
  days: number;
  resolutionIdx: number;
  modelsMask: number;
}

export function parseRequest(body: string): ForecastParams {
  const words = body.toLowerCase().trim().split(/\s+/);
  let locationIdx = 0;
  let lat: number | undefined;
  let lon: number | undefined;
  let days = 10;
  let resolutionIdx = 0;
  let modelsMask = 1; // ECMWF default

  const gpsMatch = body.match(/Lat\s+([-\d.]+)\s+Lon\s+([-\d.]+)/i);
  if (gpsMatch) {
    lat = parseFloat(gpsMatch[1]);
    lon = parseFloat(gpsMatch[2]);
  }

  for (const word of words) {
    if (word === "upper") {
      locationIdx = 0;
    } else if (word === "airstrip") {
      locationIdx = 1;
    } else if (word === "current" || word === "here") {
      locationIdx = 2;
    } else if (/^\d+d$/.test(word)) {
      days = Math.max(1, Math.min(10, parseInt(word)));
    } else if (word in RESOLUTION_LABEL_TO_IDX) {
      resolutionIdx = RESOLUTION_LABEL_TO_IDX[word];
    } else {
      const parts = word.split(",");
      if (parts.some((m) => m in MODEL_NAME_TO_BIT)) {
        let mask = 0;
        for (const m of parts) {
          if (m in MODEL_NAME_TO_BIT) mask |= 1 << MODEL_NAME_TO_BIT[m];
        }
        if (mask) modelsMask = mask;
      }
    }
  }

  return { locationIdx, lat, lon, days, resolutionIdx, modelsMask };
}

export async function fetchForecast(params: ForecastParams): Promise<string> {
  const locationNames = Object.keys(LOCATION_COORDS);
  const locationName = locationNames[params.locationIdx];
  let lat: number, lon: number, tz: string;
  if (params.locationIdx === 2) {
    if (params.lat == null || params.lon == null) throw new Error("current location requested but no GPS coordinates in message");
    [lat, lon, tz] = [params.lat, params.lon, "America/Anchorage"];
  } else {
    [lat, lon, tz] = LOCATION_COORDS[locationName];
  }

  const modelKeys = (["ECMWF", "GFS", "ICON"] as const).filter(
    (_, bit) => params.modelsMask & (1 << bit),
  );
  const keys = modelKeys.length ? modelKeys : (["ECMWF"] as const);

  const targetHours = RESOLUTION_TARGET_HOURS[params.resolutionIdx];
  const daily = params.resolutionIdx === 0;

  const rowsPerModel: Row[][] = await Promise.all(
    keys.map((key) =>
      daily
        ? noonRows(key, params.days, lat, lon, tz)
        : hourRows(key, params.days, targetHours!, lat, lon, tz),
    ),
  );

  const firstTime = rowsPerModel[0][0].time;
  const month = parseInt(firstTime.slice(5, 7));
  const day = parseInt(firstTime.slice(8, 10));
  const hour = parseInt(firstTime.slice(11, 13));

  const msg: ForecastMessage = {
    version: VERSION,
    location: params.locationIdx,
    days: params.days,
    resolution: params.resolutionIdx,
    models_mask: params.modelsMask,
    month,
    day,
    hour,
    periods: rowsPerModel.map((rows) => rows.map((r) => toFullPeriod(r, daily))),
  };

  return messageToString(msg);
}
